### PR TITLE
fixStackOverflowErrorWhenSerializingCosmosDiagnostics

### PR DESCRIPTION
**Pull request by xinlian12**
**Issue:**
Within java SDK, when SDK sending the request to server, it will also spin up a separate task which will cancel the request if SDK did not get the response within the expected time. And when SDK get the response back, this task will get cancelled accordingly. 
 
However for some rare cases, when the following combination condition happens, and overflow error can happen:

-  SDK get a response back from backend but is NotFoundException or any other type exceptions
-  The timer task suppose to be cancelled did not get cancelled properly
-  The original flow went through, and CosmosDiagnosticsContext is being set in the CosmosDiagnostics 

Because Step2 is not being cancelled properly, so the expire() flow still get triggered, and due to the request failed with exception, as part of the serializing requestRecord flow, it also tried to serialize the exception. `CosmosDiagnostics` is one of the property in `CosmosException`,  since there is no customized serializer on the CosmosDiagnostics, so it uses default serializer(which decides the property to be serialized based on the property and getter methods), and that is when it can hit the `StackOverflowError`. 

`com.fasterxml.jackson.databind.JsonMappingException: Infinite recursion (StackOverflowError) (through reference chain: com.azure.cosmos.CosmosDiagnosticsContext["diagnostics"]->java.util.concurrent.ConcurrentLinkedDeque[0]->com.azure.cosmos.CosmosDiagnostics["diagnosticsContext"]->com.azure.cosmos.CosmosDiagnosticsContext["diagnostics"]->java.util.concurrent.ConcurrentLinkedDeque[0]->com.azure.cosmos.CosmosDiagnostics["diagnosticsContext"]->com.azure.cosmos.CosmosDiagnosticsContext["diagnostics"]->java.util.concurrent.ConcurrentLinkedDeque[0]->com.azure.cosmos.CosmosDiagnostics["diagnosticsContext"]->com.azure.cosmos.CosmosDiagnosticsContext["diagnostics"]->java.util.concurrent.ConcurrentLinkedDeque[0]->com.azure.cosmos.CosmosDiagnostics["diagnosticsContext"]->com.azure.cosmos.CosmosDiagnosticsContext["diagnostics"]->java.util.concurrent.ConcurrentLinkedDeque[0]->com.azure.cosmos.CosmosDiagnostics["diagnosticsContext"]->com.azure.cosmos.CosmosDiagnosticsContext["di`

**Fixes:**
1. In order to break the circular reference, added `JsonIgnore` on `getDiagnosticsContext` within `CosmosDiagnostics`
2. In `RntbdRequestRecord`, changed to use error.toString when serializing the error

### CODE REVIEW

File: sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestRecordTests.java [hunk 1]
```diff
@@ -90,4 +99,52 @@ public void cancelRecord() throws URISyntaxException, InterruptedException, Json
         assertThat(errorStatus).isNotNull();
         assertThat(errorStatus.toString()).isEqualTo(statusString);
     }
+
+    @Test(groups = { "fast" })
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestRecordTests.java:103]: Mixing Unit tests and fast tests is not a great idea. Now for the unit tests as well, you will initialize TestSuiteBase class for this particular class unnecessary. 
Can we keep them separate?

File: sdk/cosmos/azure-cosmos/CHANGELOG.md [hunk 2]
```diff
@@ -7,13 +7,14 @@
 #### Breaking Changes
 
 #### Bugs Fixed
+* Fixed an issue where `StackOverflowError` can happen when using default `ObjectMapper` to serialize `CosmosDiagnostics` - See[PR 40272](https://github.com/Azure/azure-sdk-for-java/pull/40272)
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/CHANGELOG.md:10]:
```suggestion
* Fixed a rare issue causing `StackOverflowError` when `RntbdRequestRecord` expires and tries to serialize `CosmosException` using default Jackson Object Mapper - See[PR 40272](https://github.com/Azure/azure-sdk-for-java/pull/40272)
```

**Review by kushagraThapar — APPROVED** [review]
LGTM, thanks @xinlian12

File: sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestRecordTests.java [hunk 3]
```diff
@@ -90,4 +99,52 @@ public void cancelRecord() throws URISyntaxException, InterruptedException, Json
...
+            RntbdRequestTimer requestTimer = new RntbdRequestTimer(5000, 5000);
+            RntbdRequestRecord record = new AsyncRntbdRequestRecord(requestArgs, requestTimer);
+            record.setSendingRequestHasStarted();
+            record.completeExceptionally(exception);
+            // validate record.toString() will work correctly
+            String recordString = record.toString();
+            assertThat(recordString.contains("NotFoundException")).isTrue();
+        } finally {
+            if (client != null) {
+                client.close();
```
Comment (jeet1995) [inline sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestRecordTests.java:146]: nit: Can we use `safeClose()`?

**Review by jeet1995 — APPROVED** [review]
LGTM - thanks!

File: sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestRecordTests.java (hunk 1 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestRecordTests.java:103]: where is the problem with initializing TestSuiteBase?

Comment (xinlian12) [reply sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestRecordTests.java]: I moved the tests into cosmosDiagnosticsTests file as part of the test is testing the serialization of the cosmos diagnostics

File: sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestRecordTests.java (hunk 3 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos-tests/src/test/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestRecordTests.java]: moved the tests to CosmosDiagnosticsTest and changed to use safeClose()

**Comment by xinlian12** [comment]
Tested the following tests locally:
readAllItemsWithDetailMetricsWithExplicitPageSize
readItem

**Comment by xinlian12** [comment]
the failed tests does not related to the change in this PR

**Pull request merged**